File: python/airforce/resources/notifications.py
"""Notification preferences, the in-app feed, and delivery-channel linking."""

from __future__ import annotations

from typing import Any, Dict, List, Optional

from ._base import AsyncAPIResource, SyncAPIResource, clean, enc

_SESSION = "session"


class Notifications(SyncAPIResource):
    def get_prefs(self, **kw: Any) -> Any:
        return self._transport.request("GET", "/api/me/notification-prefs", auth=_SESSION, **kw)

    def update_prefs(self, prefs: Dict[str, Any], **kw: Any) -> Any:
        """Partially update notification preferences. The dict is sent verbatim:
        absent fields stay unchanged and ``{"quiet_hours": None}`` clears quiet
        hours. Unknown channel ids in ``routing`` are dropped server-side."""
        return self._transport.request("PATCH", "/api/me/notification-prefs", auth=_SESSION, json=prefs, **kw)

    def list(self, *, limit: Optional[int] = None, before: Optional[str] = None, **kw: Any) -> Any:
        """In-app feed, newest first → ``{items: [FeedItem], unread}``. ``limit`` is
        1-100 (default 30); ``before`` is a ``created_at`` cursor for paging."""
        return self._transport.request("GET", "/api/me/notifications", auth=_SESSION,
                                       params=clean(limit=limit, before=before), **kw)

    def mark_read(self, ids: Optional[List[str]] = None, *, mark_all: Optional[bool] = None, **kw: Any) -> Any:
        """Mark feed items read — by ``ids`` or all of them (``mark_all=True``)."""
        return self._transport.request("POST", "/api/me/notifications/read", auth=_SESSION,
                                       json=clean(ids=ids, all=mark_all), **kw)

    def channels(self, **kw: Any) -> Any:
        """Linked delivery-channel identities + linkable channel ids →
        ``{identities: [ChannelIdentity], available_channels}``."""
        return self._transport.request("GET", "/api/me/channels", auth=_SESSION, **kw)

    def link_channel(self, *, channel: str, address: str = "", display: Optional[str] = None, **kw: Any) -> Any:
        """Start linking a channel; the verification code is delivered through the
        channel itself (30-minute expiry). Bot channels accept an empty ``address``
        and return a one-time link code / deep link instead."""
        return self._transport.request("POST", "/api/me/channels", auth=_SESSION,
                                       json=clean(channel=channel, address=address, display=display), **kw)

    def verify_channel(self, *, channel: str, code: str, **kw: Any) -> Any:
        """Complete channel verification with the delivered code (400 on an invalid
        or expired code)."""
        return self._transport.request("POST", "/api/me/channels/verify", auth=_SESSION,
                                       json={"channel": channel, "code": code}, **kw)

    def unlink_channel(self, channel: str, **kw: Any) -> Any:
        return self._transport.request("DELETE", f"/api/me/channels/{enc(channel)}", auth=_SESSION, **kw)


class AsyncNotifications(AsyncAPIResource):
    async def get_prefs(self, **kw: Any) -> Any:
        return await self._transport.request("GET", "/api/me/notification-prefs", auth=_SESSION, **kw)

    async def update_prefs(self, prefs: Dict[str, Any], **kw: Any) -> Any:
        return await self._transport.request("PATCH", "/api/me/notification-prefs", auth=_SESSION, json=prefs, **kw)

    async def list(self, *, limit: Optional[int] = None, before: Optional[str] = None, **kw: Any) -> Any:
        return await self._transport.request("GET", "/api/me/notifications", auth=_SESSION,
                                             params=clean(limit=limit, before=before), **kw)

    async def mark_read(self, ids: Optional[List[str]] = None, *, mark_all: Optional[bool] = None, **kw: Any) -> Any:
        return await self._transport.request("POST", "/api/me/notifications/read", auth=_SESSION,
                                             json=clean(ids=ids, all=mark_all), **kw)

    async def channels(self, **kw: Any) -> Any:
        return await self._transport.request("GET", "/api/me/channels", auth=_SESSION, **kw)

    async def link_channel(self, *, channel: str, address: str = "", display: Optional[str] = None, **kw: Any) -> Any:
        return await self._transport.request("POST", "/api/me/channels", auth=_SESSION,
                                             json=clean(channel=channel, address=address, display=display), **kw)

    async def verify_channel(self, *, channel: str, code: str, **kw: Any) -> Any:
        return await self._transport.request("POST", "/api/me/channels/verify", auth=_SESSION,
                                             json={"channel": channel, "code": code}, **kw)

    async def unlink_channel(self, channel: str, **kw: Any) -> Any:
        return await self._transport.request("DELETE", f"/api/me/channels/{enc(channel)}", auth=_SESSION, **kw)

File: go/notifications.go
package airforce

import (
	"context"
	"net/http"
	"net/url"
	"strconv"
)

// NotificationsService accesses notification preferences, the in-app feed and
// delivery-channel linking under /api/me/*. All endpoints require a session
// token.
type NotificationsService struct{ client *Client }

// GetPrefs returns the caller's notification preferences.
func (s *NotificationsService) GetPrefs(ctx context.Context) (map[string]any, error) {
	var out map[string]any
	err := s.client.getJSON(ctx, "/api/me/notification-prefs", "session", nil, &out)
	return out, err
}

// UpdatePrefs partially updates the preferences: absent fields stay unchanged,
// "quiet_hours": nil clears quiet hours, unknown channel ids in routing are
// dropped server-side. Returns the updated preferences.
func (s *NotificationsService) UpdatePrefs(ctx context.Context, patch map[string]any) (map[string]any, error) {
	var out map[string]any
	err := s.client.postJSONMethod(ctx, http.MethodPatch, "/api/me/notification-prefs", "session", patch, &out)
	return out, err
}

// List returns the in-app feed, newest first, as {items, unread}. limit is
// 1-100 (0 = server default of 30); before is a created_at cursor for paging
// (empty = start at the newest item).
func (s *NotificationsService) List(ctx context.Context, limit int, before string) (map[string]any, error) {
	query := url.Values{}
	if limit > 0 {
		query.Set("limit", strconv.Itoa(limit))
	}
	if before != "" {
		query.Set("before", before)
	}
	if len(query) == 0 {
		query = nil
	}
	var out map[string]any
	err := s.client.getJSON(ctx, "/api/me/notifications", "session", query, &out)
	return out, err
}

// MarkRead marks feed items read by ids, or everything when all is true.
// Returns {updated, unread}.
func (s *NotificationsService) MarkRead(ctx context.Context, ids []string, all bool) (map[string]any, error) {
	body := map[string]any{}
	if len(ids) > 0 {
		body["ids"] = ids
	}
	if all {
		body["all"] = true
	}
	var out map[string]any
	err := s.client.postJSON(ctx, "/api/me/notifications/read", "session", body, &out)
	return out, err
}

// Channels lists linked delivery-channel identities plus the linkable channel
// ids ({identities, available_channels}).
func (s *NotificationsService) Channels(ctx context.Context) (map[string]any, error) {
	var out map[string]any
	err := s.client.getJSON(ctx, "/api/me/channels", "session", nil, &out)
	return out, err
}

// LinkChannel starts linking a delivery channel. The verification code is
// delivered through the channel itself (30-minute expiry); bot channels accept
// an empty address and return a one-time link code / deep link instead
// ({status: "link_ready", code, deep_link?, expires_minutes}).
func (s *NotificationsService) LinkChannel(ctx context.Context, channel, address, display string) (map[string]any, error) {
	body := map[string]any{"channel": channel, "address": address}
	if display != "" {
		body["display"] = display
	}
	var out map[string]any
	err := s.client.postJSON(ctx, "/api/me/channels", "session", body, &out)
	return out, err
}

// VerifyChannel completes channel verification with the delivered code
// (400 on an invalid or expired code).
func (s *NotificationsService) VerifyChannel(ctx context.Context, channel, code string) (map[string]any, error) {
	var out map[string]any
	err := s.client.postJSON(ctx, "/api/me/channels/verify", "session", map[string]any{"channel": channel, "code": code}, &out)
	return out, err
}

// UnlinkChannel revokes a linked channel identity.
func (s *NotificationsService) UnlinkChannel(ctx context.Context, channel string) (map[string]any, error) {
	var out map[string]any
	err := s.client.deleteJSON(ctx, "/api/me/channels/"+url.PathEscape(channel), "session", &out)
	return out, err
}
